### PR TITLE
CORDA-1392 Make Hibernate entities non-Serializable

### PR DESCRIPTION
**Pull request by tudor-malene**
See: https://r3-cev.atlassian.net/browse/CORDA-1392

The problem seems to be that in some circumstances Hibernate "deep copies" objects by serialising/deserialising them. But only if they are Serialisable.

This breaks as we disabled java serialisation.

The obvious fix is to remove "Serialisable" from our Jpa entities ( It is not required by JPA or hibernate).

This should convince Hibernate to no longer use the "SerializableTypeDescriptor".

The only gotcha is that Hibernate requires that "PersistentStateRef" is Serialisable as it's a composite-id. ( and that's a requirement because hibernate expects all Id objects to be serialisable as it might use them as keys on distributed caches). 
The fix is to mark that class as "@Immutable", and it will not try to clone it any more.

Other changes:
 - StatePersistable is part of the public api and it was serialisable. As @mikehearn  already mentioned that is an ABI break, but we can live with it.
 - Immutable was added to a DJVM class, so ProGuard was instructed to ignore it.

### CODE REVIEW

**Comment by mikehearn** [comment]
It's technically going to be an ABI compatibility break which is why the API checker cries. However I very much doubt it will affect anyone and it's the best fix for this bug, so I'll OK it this time, as long as we mention it in the changelog.

File: core/src/main/kotlin/net/corda/core/schemas/PersistentTypes.kt [hunk 1]
```diff
@@ -82,6 +83,7 @@ class PersistentState(@EmbeddedId var stateRef: PersistentStateRef? = null) : St
  */
 @KeepForDJVM
 @Embeddable
+@Immutable
 data class PersistentStateRef(
```
Comment (chrisr3) [inline core/src/main/kotlin/net/corda/core/schemas/PersistentTypes.kt]: Do we need this Hibernate-specific `@Immutable` annotation? Can't we just use
```kotlin
@Column(..., updatable = false, ...)
```
for both of these fields instead?

Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/schemas/PersistentTypes.kt]: Unfortunately this stuff is buried deep in Hibernate.

And, it depends on this  specific hibernate annotation.

See: org.hibernate.type.descriptor.java.SerializableTypeDescriptor.createMutabilityPlan

File: docs/source/changelog.rst [hunk 2]
```diff
@@ -6,6 +6,9 @@ release, see :doc:`upgrade-notes`.
 
 Unreleased
 ----------
+* The Corda JPA entities are no longer "java.io.Serializable". This is a mild binary compatibility break, but affects only clients who relied on this fact.
```
Comment (mikehearn) [inline docs/source/changelog.rst]: I guess it's obvious that it only affects clients who relied on it :-) How about this wording:

The Corda JPA entities no longer implement `java.io.Serializable`, as this was causing persistence errors in obscure cases. Java serialization is disabled globally in the node, but in the unlikely event you were relying on these types being Java serializable please contact us.
